internal/handlers/pg/pgdb/query.go

```diff
@@ -268,46 +269,81 @@ func prepareWhereClause(sqlFilters *types.Document) (string, []any, error) {
-		if k == "_id" {
-			switch v := v.(type) {
-			case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
-				// type not supported for pushdown
-				// TODO $eq and $ne https://github.com/FerretDB/FerretDB/issues/1840
-				// TODO $gt and $lt https://github.com/FerretDB/FerretDB/issues/1875
-			case float64, string, types.ObjectID, int32, int64:
-				// Select if value under the key is equal to provided value.
-				sql := `(_jsonb->%[1]s)::jsonb = %[2]s`
-
-				// placeholder p.Next() returns SQL argument references such as $1, $2 to prevent SQL injections.
-				// placeholder $1 is used for field key,
-				// placeholder $2 is used for field value v.
-				filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next()))
-				args = append(args, k, string(must.NotFail(pjson.MarshalSingleValue(v))))
-
-			default:
-				panic(fmt.Sprintf("Unexpected type of value: %v", v))
-			}
+		// check if values are equal or if the left value contains the right one
+		eqOperator := "@>"
 
-			continue
+		if k == "_id" {
+			// check if values are equal
+			eqOperator = "="
 		}
 
 		switch v := v.(type) {
-		case *types.Document, *types.Array, types.Binary, bool, time.Time, types.NullType, types.Regex, types.Timestamp:
+		case *types.Document:
+			iter := v.Iterator()
+			defer iter.Close()
+
+			for {
+				docKey, docVal, err := iter.Next()
+				if err != nil {
+					if errors.Is(err, iterator.ErrIteratorDone) {
+						break
+					}
+
+					return "", nil, lazyerrors.Error(err)
+				}
+
+				switch docKey {
+				case "$eq":
+					switch v := docVal.(type) {
+					case *types.Document, *types.Array, types.Binary, bool,
+						time.Time, types.NullType, types.Regex, types.Timestamp:
+						// type not supported for pushdown
+					case float64, string, types.ObjectID, int32, int64:
+						// Select if value under the key is equal to provided value.
+						sql := `(_jsonb->%[1]s)::jsonb ` + eqOperator + ` %[2]s`
+
+						filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next()))
+						args = append(args, k, string(must.NotFail(pjson.MarshalSingleValue(v))))
+					default:
+						panic(fmt.Sprintf("Unexpected type of value: %v", v))
+					}
+
+				case "$ne":
+					switch v := docVal.(type) {
+					case *types.Document, *types.Array, types.Binary, bool,
+						time.Time, types.NullType, types.Regex, types.Timestamp:
+						// type not supported for pushdown
+					case float64, string, types.ObjectID, int32, int64:
+						sql := `NOT ( ` +
+							// does document contain the key,
+							// it is necessary, as NOT won't work correctly if the key does not exist.
+							`_jsonb ? %[1]s AND ` +
+							// does the value under the key is equal to the filter
+							`(_jsonb->%[1]s)::jsonb ` + eqOperator + ` %[2]s AND ` +
+							// does the value type is equal to the filter's one
+							`(_jsonb->'$s'->'p'->%[1]s->'t')::jsonb = '"` + pjson.GetTypeOfValue(v) +
```

Comment:
Why do we check type only for `$ne`, but not for `$eq`?

Reply:
If you run `{v:{$eq:42}}` against the collection that contains int(42) and Timestamp(42), it's okay to prefetch both of them, as we filter those on our side.

For `$ne` we can't apply the same logic. If we don't check for type for `{v:{$ne:42}`, PostgreSQL will match both int and Timestamp documents, and invert results, so both of them won't be passed to FerretDB.

Reply:
```
	--- FAIL: TestQueryCompatRunner/NeDoubleZero/TestQueryCompatRunner_DateTimes (0.02s)
		query_compat_test.go:193: Compat (expected) IDs: [datetime datetime-epoch datetime-null datetime-year-max datetime-year-min]
		query_compat_test.go:193: Target (actual)   IDs: [datetime datetime-null datetime-year-max datetime-year-min]
	--- FAIL: TestQueryCompatRunner/NeDoubleZero/TestQueryCompatRunner_Scalars (0.04s)
		query_compat_test.go:193: Compat (expected) IDs: [binary binary-empty bool-false bool-true datetime datetime-epoch datetime-year-max datetime-year-min double double-1 double-2 double-3 double-4 double-5 double-6 double-7 double-big double-max double-max-overflow double-max-overflow-verge double-min-overflow double-min-overflow-verge double-smallest double-whole int32 int32-1 int32-2 int32-3 int32-max int32-min int64 int64-1 int64-2 int64-3 int64-big int64-max int64-min null objectid objectid-empty regex regex-empty string string-double string-empty string-whole timestamp timestamp-i unset]
		query_compat_test.go:193: Target (actual)   IDs: [binary binary-empty bool-false bool-true datetime datetime-year-max datetime-year-min double double-1 double-2 double-3 double-4 double-5 double-6 double-7 double-big double-max double-max-overflow double-max-overflow-verge double-min-overflow double-min-overflow-verge double-smallest double-whole int32 int32-1 int32-2 int32-3 int32-max int32-min int64 int64-1 int64-2 int64-3 int64-big int64-max int64-min null objectid objectid-empty regex regex-empty string string-double string-empty string-whole timestamp timestamp-i unset]
	--- FAIL: TestQueryCompatRunner/NeInt64Zero/TestQueryCompatRunner_DateTimes (0.11s)
		query_compat_test.go:193: Compat (expected) IDs: [datetime datetime-epoch datetime-null datetime-year-max datetime-year-min]
		query_compat_test.go:193: Target (actual)   IDs: [datetime datetime-null datetime-year-max datetime-year-min]
	--- FAIL: TestQueryCompatRunner/NeInt64Zero/TestQueryCompatRunner_Scalars (0.04s)
		query_compat_test.go:193: Compat (expected) IDs: [binary binary-empty bool-false bool-true datetime datetime-epoch datetime-year-max datetime-year-min double double-1 double-2 double-3 double-4 double-5 double-6 double-7 double-big double-max double-max-overflow double-max-overflow-verge double-min-overflow double-min-overflow-verge double-smallest double-whole int32 int32-1 int32-2 int32-3 int32-max int32-min int64 int64-1 int64-2 int64-3 int64-big int64-max int64-min null objectid objectid-empty regex regex-empty string string-double string-empty string-whole timestamp timestamp-i unset]
		query_compat_test.go:193: Target (actual)   IDs: [binary binary-empty bool-false bool-true datetime datetime-year-max datetime-year-min double double-1 double-2 double-3 double-4 double-5 double-6 double-7 double-big double-max double-max-overflow double-max-overflow-verge double-min-overflow double-min-overflow-verge double-smallest double-whole int32 int32-1 int32-2 int32-3 int32-max int32-min int64 int64-1 int64-2 int64-3 int64-big int64-max int64-min null objectid objectid-empty regex regex-empty string string-double string-empty string-whole timestamp timestamp-i unset]
	--- FAIL: TestQueryCompatRunner/NeString/TestQueryCompatRunner_Regexes (0.02s)
		query_compat_test.go:193: Compat (expected) IDs: [regex regex-empty regex-null]
		query_compat_test.go:193: Target (actual)   IDs: [regex-empty regex-null]
	--- FAIL: TestQueryCompatRunner/NeString/TestQueryCompatRunner_ArrayRegexes (0.02s)
		query_compat_test.go:193: Compat (expected) IDs: [array-regex]
		query_compat_test.go:193: Target (actual)   IDs: []
```